File: pb/src/main/java/org/sqlbuilder/sl/SlModule.java
package org.sqlbuilder.sl;

import org.sqlbuilder.common.Module;
import org.sqlbuilder.pb.Inserter;
import org.sqlbuilder.sl.collectors.SemlinkProcessor;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.Properties;

public class SlModule extends Module
{
	public static final String MODULE_ID = "sl";

	protected SlModule(final String conf, final Mode mode)
	{
		super(MODULE_ID, conf, mode);
	}

	@Override
	protected void run()
	{
		try
		{
			buildSemLink(props);
		}
		catch (IOException e)
		{
			e.printStackTrace();
		}
	}


	private static void buildSemLink(final Properties props) throws FileNotFoundException
	{
		new SemlinkProcessor(props).run();
	}

	public static void main(final String[] args)
	{
		int i = 0;
		Mode mode = Mode.PLAIN;
		if (args[i].startsWith("-"))
		{
			mode = Mode.read(args[i++]);
		}
		String conf = args[i];
		new SlModule(conf, mode).run();
	}
}

File: sl/src/main/java/org/semantikos/sl/SlModule.kt
package org.semantikos.sl

import org.semantikos.common.Module
import org.semantikos.common.Module.Mode.Companion.read
import org.semantikos.sl.collectors.SemlinkProcessor
import org.semantikos.sl.collectors.SemlinkUpdatingProcessor
import java.io.IOException

class SlModule(
    conf: String,
    mode: Mode
) : Module(MODULE_ID, conf, mode) {

    override fun run() {
        checkNotNull(props)

        when (mode) {
            Mode.PLAIN, Mode.RESOLVE -> {
                SemlinkProcessor(props).run()
                try {
                    val inserter = if (mode == Mode.PLAIN) Inserter(props) else ResolvingInserter(props)
                    inserter.insert()
                } catch (e: IOException) {
                    e.printStackTrace()
                } catch (e: ClassNotFoundException) {
                    e.printStackTrace()
                }
            }

            Mode.UPDATE              -> {
                SemlinkUpdatingProcessor(props).run()
                try {
                    val inserter: Inserter = ResolvingUpdater(props)
                    inserter.insert()
                } catch (e: IOException) {
                    e.printStackTrace()
                } catch (e: ClassNotFoundException) {
                    e.printStackTrace()
                }
            }

            Mode.EXPORT              -> {}
            else                     -> {}
        }
    }

    companion object {

        const val MODULE_ID: String = "sl"

        @JvmStatic
        fun main(args: Array<String>) {
            var i = 0
            var mode = Mode.PLAIN
            if (args[i].startsWith("-")) {
                mode = read(args[i++])
            }
            val conf = args[i]
            SlModule(conf, mode).run()
        }
    }
}
